docs/Execution.md

```diff
@@ -0,0 +1,107 @@
+Execution
+======
+
+The MsQuic API uses a different execution model compared to BSD-style sockets and most other networking libraries built on them.
+The sections below detail the designs MsQuic uses and the reasons behind these choices.
+
+## Event Model
+
+In the MsQuic API, all state changes and notifications are indicated directly to the application via a callback.
+This includes connection state changes, new streams being created, stream data being received, and stream sends completing.
+
+```c
+typedef struct QUIC_LISTENER_EVENT {
+    QUIC_LISTENER_EVENT_TYPE Type;
+    union {
+        struct { ... } NEW_CONNECTION;
+        struct { ... } STOP_COMPLETE;
+    };
+} QUIC_LISTENER_EVENT;
+
+typedef
+_IRQL_requires_max_(PASSIVE_LEVEL)
+_Function_class_(QUIC_LISTENER_CALLBACK)
+QUIC_STATUS
+(QUIC_API QUIC_LISTENER_CALLBACK)(
+    _In_ HQUIC Listener,
+    _In_opt_ void* Context,
+    _Inout_ QUIC_LISTENER_EVENT* Event
+    );
+```
+
+Above is an example of a callback delivered to the listener interface.
+The application must register a callback handler to manage all events MsQuic may indicate, returning a status to show if it was successfully handled or not.
+
+This approach differs significantly from BSD sockets, where the application must make a call (e.g., `send` or `recv`) to determine if something happened.
```

Comment:
it would also be useful to compare this to Winsock OVERLAPPED I/O APIs. since this seems like it's a combination of "OVERLAPPED I/O semantics + select/event semantics". e.g., when I connect my OVERLAPPED I/O request to an NT Threadpool, or drain completions manually from an IO completion port, things start to look very similar.

It *sounds* so far like the difference is that QUIC gives these callback invocations on state changes as well - not just posting IO -- which to me sounds much more like the select/event model, when the app is notified when the underlying socket has changed some internal state ... (e.g. "read to send data now" state, or "ready to receive data now" state, or "your connect request completed" state, or "you have new inbound connections ready to accept" state, or "your socket is in a failed state which we RST" state)

Reply:
QUIC is fundamentally different. You set up your object, be it a listener, connection or stream, up front with a callback and your context pointer **once**. Then all events happen for that object for the rest of its lifetime.